### PR TITLE
Added dynamic authentication with external command

### PR DESCRIPTION
**Pull request by levkohimins**
## Description

[Create configuration for Terragrunt to request dynamic authentication configurations at runtime
](https://linear.app/gruntwork/issue/SME-853/create-configuration-for-terragrunt-to-request-dynamic-authentication)

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [ ] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [ ] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added / Removed / Updated [X].

### CODE REVIEW

File: cli/commands/terraform/helper/aws_credentials.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+	opts.Env["AWS_ACCESS_KEY_ID"] = aws.StringValue(creds.AccessKeyId)
+	opts.Env["AWS_SECRET_ACCESS_KEY"] = aws.StringValue(creds.SecretAccessKey)
+	opts.Env["AWS_SESSION_TOKEN"] = aws.StringValue(creds.SessionToken)
+	opts.Env["AWS_SECURITY_TOKEN"] = aws.StringValue(creds.SessionToken)
+
+	return nil
+}
+
+func getAWSCredentialsByRunningProviderCommand(ctx context.Context, opts *options.TerragruntOptions) (*sts.Credentials, error) {
```
Comment (yhakbar) [inline cli/commands/terraform/helper/aws_credentials.go:47]: This capability is more generic than getting AWS creds, @levkohimins .

Users can populate whatever environment variables they want with this command, including other clouds or services.

e.g. Someone could dynamically fetch CloudFlare credentials using this feature.

The way in which this is especially useful in the context of AWS is that someone could dynamically determine which role they want to assume, and use different credentials for accessing it on the fly this way.

File: cli/commands/flags.go [hunk 2]
```diff
@@ -309,6 +312,12 @@ func NewGlobalFlags(opts *options.TerragruntOptions) cli.Flags {
 			EnvVar:      TerragruntProviderCacheRegistryNamesEnvVarName,
 			Usage:       "The list of remote registries to cached by Terragrunt Provider Cache server. By default, 'registry.terraform.io', 'registry.opentofu.org'.",
 		},
+		&cli.GenericFlag[string]{
+			Name:        TerragruntAuthProviderCmdFlagName,
+			Destination: &opts.AuthProviderCmd,
+			EnvVar:      TerragruntAuthProviderCmdEnvVarName,
+			Usage:       "Path to the authentication command to obtain AWS credentials.",
```
Comment (yhakbar) [inline cli/commands/flags.go:319]: More generic than that.

File: cli/commands/terraform/helper/aws_credentials.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	output, err := shell.RunShellCommandWithOutput(ctx, opts, "", true, false, command, args...)
+	if err != nil {
+		return nil, err
+	}
+
+	if output.Stdout == "" {
+		return nil, errors.Errorf("the command %s completed successfully, but the response does not contain json string", command)
+	}
+
+	type credentials struct {
```
Comment (yhakbar) [inline cli/commands/terraform/helper/aws_credentials.go:68]: You can parse the response as a `map[string]string` and set the environment variables to what it tells you. This shouldn't have any special awareness of the fact that it might be used for AWS.

File: docs/_docs/04_reference/cli-options.md [hunk 4]
```diff
@@ -1306,3 +1306,21 @@ Specify the plan output directory for the `*-all` commands. Useful to save plans
...
 
 Specify the output directory for the `*-all` commands to store plans in JSON format. Useful to read plans programmatically.
+
+### terragrunt-auth-provider-cmd
+
+**CLI Arg**: `--terragrunt-auth-provider-cmd`<br/>
+**Environment Variable**: `TERRAGRUNT_AUTH_PROVIDER_CMD`<br/>
+**Requires an argument**: `--terragrunt-auth-provider-cmd /path/to/executable/file`<br/>
+
+The path to the executable file to obtain authentication credentials. If specified, Terragrunt runs this command for every working directory. The output must be the following JSON string
```
Comment (yhakbar) [inline docs/_docs/04_reference/cli-options.md:1316]:
```suggestion
The path to the executable file to obtain authentication credentials. If specified, Terragrunt runs this command for every working directory.

The output must be valid JSON of the following schema:
```

Then show that it's a map of string to string, and include some keys that are not related to AWS.

File: test/fixture-auth-provider-cmd/app1/creds.config [hunk 5]
```diff
@@ -0,0 +1,3 @@
+fake_id=app1-ASIA6ODU3VLNW2YGRZ12
```
Comment (yhakbar) [inline test/fixture-auth-provider-cmd/app1/creds.config:1]: We can leave these as a lot less real so that it doesn't scare the secrets scanner 😄 

It can be `fake_id=fake_id`, etc.

File: test/integration_test.go [hunk 6]
```diff
@@ -4123,6 +4124,17 @@ func TestReadTerragruntConfigIamRole(t *testing.T) {
...
 
+func TestReadTerragruntAuthProviderCmd(t *testing.T) {
+	t.Parallel()
+
+	cleanupTerraformFolder(t, TEST_FIXTURE_AUTH_PROVIDER_CMD)
+	tmpEnvPath := copyEnvironment(t, TEST_FIXTURE_AUTH_PROVIDER_CMD)
+	rootPath := util.JoinPath(tmpEnvPath, TEST_FIXTURE_AUTH_PROVIDER_CMD)
+	mockPiplinesCmd := filepath.Join(rootPath, "mock-pipelines.sh")
+
+	runTerragrunt(t, fmt.Sprintf("terragrunt run-all init --terragrunt-working-dir %s --terragrunt-auth-provider-cmd %s", rootPath, mockPiplinesCmd))
```
Comment (yhakbar) [inline test/integration_test.go:4135]: If you're looking for a way to prove that this is working, you can just output the following from the script:

```json
{
  "envs": {
    "TF_VAR_foo": "bar"
  }
}
```

And have a `variable` `foo`, then read it in the outputs.

All we need to prove is that we can set some environment variables before the call to OpenTofu/Terraform is done.

File: docs/_docs/04_reference/cli-options.md (hunk 4 above)
Comment (levkohimins) [reply docs/_docs/04_reference/cli-options.md]: Updated, thanks!

File: test/fixture-auth-provider-cmd/app1/creds.config (hunk 5 above)
Comment (levkohimins) [reply test/fixture-auth-provider-cmd/app1/creds.config]: Fixed, thanks)

File: test/integration_test.go (hunk 6 above)
Comment (levkohimins) [reply test/integration_test.go:4135]: Considering that we have decided to use a strictly defined structure,  I have to use `{awsCredentials: {ACCESS_KEY_ID: ....` in the test

File: cli/commands/terraform/creds/providers/externalcmd/provider.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+	output, err := shell.RunShellCommandWithOutput(ctx, provider.terragruntOptions, "", true, false, command, args...)
+	if err != nil {
+		return nil, err
+	}
+
+	if output.Stdout == "" {
+		return nil, errors.Errorf("the command %s completed successfully, but the response does not contain json string", command)
+	}
+
+	type response struct {
```
Comment (yhakbar) [inline cli/commands/terraform/creds/providers/externalcmd/provider.go:55]: Do you mind exporting this? Users creating external commands built in Golang can use this as the contract for the IPC.

File: cli/commands/terraform/creds/providers/externalcmd/provider.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+type AWSCredentials struct {
+	AccessKeyID     string `json:"ACCESS_KEY_ID"`
+	SecretAccessKey string `json:"SECRET_ACCESS_KEY"`
+	SessionToken    string `json:"SESSION_TOKEN"`
+}
+
+func (creds *AWSCredentials) Envs(opts *options.TerragruntOptions) map[string]string {
+	var emtpyFields []string
```
Comment (yhakbar) [inline cli/commands/terraform/creds/providers/externalcmd/provider.go:86]:
```suggestion
	var emptyFields []string
```

File: cli/commands/terraform/creds/providers/externalcmd/provider.go [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	type response struct {
+		AWSCredentials *AWSCredentials   `json:"awsCredentials"`
+		Envs           map[string]string `json:"envs"`
+	}
+	resp := &response{Envs: make(map[string]string)}
+
+	if err := json.Unmarshal([]byte(output.Stdout), &resp); err != nil {
+		return nil, errors.WithStackTrace(err)
```
Comment (yhakbar) [inline cli/commands/terraform/creds/providers/externalcmd/provider.go:62]: I think this should be a more generic error in order to avoid accidentally leaking credentials.

File: cli/commands/terraform/creds/providers/externalcmd/provider.go [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	creds := &providers.Credentials{
+		Name: providers.AWSCredentials,
+		Envs: resp.Envs,
+	}
+
+	if resp.AWSCredentials != nil {
+		if envs := resp.AWSCredentials.Envs(provider.terragruntOptions); envs != nil {
+			maps.Copy(creds.Envs, envs)
```
Comment (yhakbar) [inline cli/commands/terraform/creds/providers/externalcmd/provider.go:72]: We should be very explicit in documentation that the same key being present in `providers.AWSCredentials` and `resp.Envs` will result in the `providers.AWSCredentials` taking precedence.

File: cli/commands/terraform/creds/providers/externalcmd/provider.go [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+	AccessKeyID     string `json:"ACCESS_KEY_ID"`
+	SecretAccessKey string `json:"SECRET_ACCESS_KEY"`
+	SessionToken    string `json:"SESSION_TOKEN"`
+}
+
+func (creds *AWSCredentials) Envs(opts *options.TerragruntOptions) map[string]string {
+	var emtpyFields []string
+
+	if creds.AccessKeyID == "" {
+		emtpyFields = append(emtpyFields, "ACCESS_KEY_ID")
```
Comment (yhakbar) [inline cli/commands/terraform/creds/providers/externalcmd/provider.go:89]:
```suggestion
		emptyFields = append(emptyFields, "ACCESS_KEY_ID")
```

File: cli/commands/terraform/creds/providers/externalcmd/provider.go [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+	}
+	if creds.SecretAccessKey == "" {
+		emtpyFields = append(emtpyFields, "SECRET_ACCESS_KEY")
```
Comment (yhakbar) [inline cli/commands/terraform/creds/providers/externalcmd/provider.go:92]:
```suggestion
		emptyFields = append(emptyFields, "SECRET_ACCESS_KEY")
```

File: cli/commands/terraform/creds/providers/externalcmd/provider.go [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+	}
+	if creds.SessionToken == "" {
+		emtpyFields = append(emtpyFields, "SESSION_TOKEN")
```
Comment (yhakbar) [inline cli/commands/terraform/creds/providers/externalcmd/provider.go:95]:
```suggestion
		emptyFields = append(emptyFields, "SESSION_TOKEN")
```

File: cli/commands/terraform/creds/providers/externalcmd/provider.go [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	if len(emtpyFields) > 0 {
```
Comment (yhakbar) [inline cli/commands/terraform/creds/providers/externalcmd/provider.go:98]:
```suggestion
	if len(emptyFields) > 0 {
```

File: cli/commands/terraform/creds/providers/externalcmd/provider.go [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+func (creds *AWSCredentials) Envs(opts *options.TerragruntOptions) map[string]string {
+	var emtpyFields []string
+
+	if creds.AccessKeyID == "" {
+		emtpyFields = append(emtpyFields, "ACCESS_KEY_ID")
+	}
+	if creds.SecretAccessKey == "" {
+		emtpyFields = append(emtpyFields, "SECRET_ACCESS_KEY")
+	}
+	if creds.SessionToken == "" {
```
Comment (yhakbar) [inline cli/commands/terraform/creds/providers/externalcmd/provider.go:94]: Not providing a session token is valid for authentication with AWS if a user is using a static key + secret pair.

File: cli/commands/terraform/creds/providers/externalcmd/provider.go [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+
+	if len(emtpyFields) > 0 {
+		opts.Logger.Warnf("The command %s completed successfully, but AWS credentials contains empty values: %s, nothing is being done.", opts.AuthProviderCmd, strings.Join(emtpyFields, ", "))
+		return nil
+	}
+
+	envs := map[string]string{
+		"AWS_ACCESS_KEY_ID":     creds.AccessKeyID,
+		"AWS_SECRET_ACCESS_KEY": creds.SecretAccessKey,
+		"AWS_SESSION_TOKEN":     creds.SessionToken,
```
Comment (yhakbar) [inline cli/commands/terraform/creds/providers/externalcmd/provider.go:106]: We might not be able to set these two if the `sessionToken` is missing.

File: docs/_docs/04_reference/cli-options.md [hunk 17]
```diff
@@ -1306,3 +1306,23 @@ Specify the plan output directory for the `*-all` commands. Useful to save plans
 - [run-all](#run-all)
 
 Specify the output directory for the `*-all` commands to store plans in JSON format. Useful to read plans programmatically.
+
+### terragrunt-auth-provider-cmd
+
+**CLI Arg**: `--terragrunt-auth-provider-cmd`<br/>
+**Environment Variable**: `TERRAGRUNT_AUTH_PROVIDER_CMD`<br/>
+**Requires an argument**: `--terragrunt-auth-provider-cmd /path/to/executable/file`<br/>
```
Comment (yhakbar) [inline docs/_docs/04_reference/cli-options.md:1314]:
```suggestion
**Requires an argument**: `--terragrunt-auth-provider-cmd "command [arguments]"`<br/>
```

File: docs/_docs/04_reference/cli-options.md [hunk 18]
```diff
@@ -1306,3 +1306,23 @@ Specify the plan output directory for the `*-all` commands. Useful to save plans
...
+
+The path to the executable file to obtain authentication credentials. If specified, Terragrunt runs this command for every working directory.
```
Comment (yhakbar) [inline docs/_docs/04_reference/cli-options.md:1316]:
```suggestion
The command and arguments used to obtain authentication credentials dynamically. If specified, Terragrunt runs this command for every working directory before running the underlying IAC for a `terragrunt.hcl` file.
```

File: cli/commands/terraform/creds/providers/externalcmd/provider.go [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+	}
+	if creds.SecretAccessKey == "" {
+		emtpyFields = append(emtpyFields, "SECRET_ACCESS_KEY")
+	}
+	if creds.SessionToken == "" {
+		emtpyFields = append(emtpyFields, "SESSION_TOKEN")
+	}
+
+	if len(emtpyFields) > 0 {
+		opts.Logger.Warnf("The command %s completed successfully, but AWS credentials contains empty values: %s, nothing is being done.", opts.AuthProviderCmd, strings.Join(emtpyFields, ", "))
```
Comment (yhakbar) [inline cli/commands/terraform/creds/providers/externalcmd/provider.go:99]:
```suggestion
		opts.Logger.Warnf("The command %s completed successfully, but AWS credentials contains empty required values: %s, nothing is being done.", opts.AuthProviderCmd, strings.Join(emtpyFields, ", "))
```

File: docs/_docs/04_reference/cli-options.md [hunk 20]
```diff
@@ -1306,3 +1306,23 @@ Specify the plan output directory for the `*-all` commands. Useful to save plans
...
+
+The output must be valid JSON of the following schema:
+
+```json
+{
```
Comment (denis256) [inline docs/_docs/04_reference/cli-options.md:1321]: Maybe make sense to add in example other non `awsCredentials` variables?

File: test/integration_test.go [hunk 21]
```diff
@@ -4123,6 +4124,28 @@ func TestReadTerragruntConfigIamRole(t *testing.T) {
...
+
+	cleanupTerraformFolder(t, TEST_FIXTURE_AUTH_PROVIDER_CMD)
+	tmpEnvPath := copyEnvironment(t, TEST_FIXTURE_AUTH_PROVIDER_CMD)
+	rootPath := util.JoinPath(tmpEnvPath, TEST_FIXTURE_AUTH_PROVIDER_CMD)
+	appPath := util.JoinPath(rootPath, "app1")
+	mockPiplinesCmd := filepath.Join(rootPath, "mock-pipelines.sh")
+
+	runTerragrunt(t, fmt.Sprintf("terragrunt run-all apply -auto-approve --terragrunt-non-interactive --terragrunt-working-dir %s --terragrunt-auth-provider-cmd %s", rootPath, mockPiplinesCmd))
+
+	stdout, _, err := runTerragruntCommandWithOutput(t, fmt.Sprintf("terragrunt output -json --terragrunt-working-dir %s --terragrunt-auth-provider-cmd %s", appPath, mockPiplinesCmd))
```
Comment (yhakbar) [inline test/integration_test.go:4138]:
```suggestion
	stdout, _, err := runTerragruntCommandWithOutput(t, fmt.Sprintf("terragrunt output -json --terragrunt-working-dir %s --terragrunt-auth-provider-cmd %s", appPath, mockAuthCmd))
```

**Review by yhakbar — CHANGES_REQUESTED** [review]

File: docs/_docs/04_reference/cli-options.md (hunk 20 above)
Comment (levkohimins) [reply docs/_docs/04_reference/cli-options.md:1321]: Agree, added. Thanks.

File: cli/commands/terraform/creds/providers/externalcmd/provider.go (hunk 7 above)
Comment (levkohimins) [reply cli/commands/terraform/creds/providers/externalcmd/provider.go]: Fair point! I did it to be able to export.

File: cli/commands/terraform/creds/providers/externalcmd/provider.go (hunk 8 above)
Comment (levkohimins) [reply cli/commands/terraform/creds/providers/externalcmd/provider.go]: Thanks :)

File: cli/commands/terraform/creds/providers/externalcmd/provider.go (hunk 9 above)
Comment (levkohimins) [reply cli/commands/terraform/creds/providers/externalcmd/provider.go]: Indeed, fixed.

File: cli/commands/terraform/creds/providers/externalcmd/provider.go (hunk 10 above)
Comment (levkohimins) [reply cli/commands/terraform/creds/providers/externalcmd/provider.go:68]: Agree. Thanks for the doc suggestion!

File: cli/commands/terraform/creds/providers/externalcmd/provider.go (hunk 12 above)
Comment (levkohimins) [reply cli/commands/terraform/creds/providers/externalcmd/provider.go]: Fixed :)

File: cli/commands/terraform/creds/providers/externalcmd/provider.go (hunk 13 above)
Comment (levkohimins) [reply cli/commands/terraform/creds/providers/externalcmd/provider.go]: Fixed :)

File: cli/commands/terraform/creds/providers/externalcmd/provider.go (hunk 14 above)
Comment (levkohimins) [reply cli/commands/terraform/creds/providers/externalcmd/provider.go]: Fixed

File: cli/commands/terraform/creds/providers/externalcmd/provider.go (hunk 15 above)
Comment (levkohimins) [reply cli/commands/terraform/creds/providers/externalcmd/provider.go]: Ah, got it. Fixed.

File: cli/commands/terraform/creds/providers/externalcmd/provider.go (hunk 16 above)
Comment (levkohimins) [reply cli/commands/terraform/creds/providers/externalcmd/provider.go:104]: Yeah, fixed.

File: docs/_docs/04_reference/cli-options.md (hunk 17 above)
Comment (levkohimins) [reply docs/_docs/04_reference/cli-options.md:1314]: Much better👍 Thanks.

File: docs/_docs/04_reference/cli-options.md (hunk 18 above)
Comment (levkohimins) [reply docs/_docs/04_reference/cli-options.md:1316]: Thanks!

File: cli/commands/terraform/creds/providers/externalcmd/provider.go (hunk 19 above)
Comment (levkohimins) [reply cli/commands/terraform/creds/providers/externalcmd/provider.go]: Nice improvement👍

File: cli/commands/terraform/creds/providers/provider.go [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+
+type CredentialsName string
+
+type Credentials struct {
+	Name CredentialsName
+	Envs map[string]string
+}
+
+type Provider interface {
+	// Name returns provdier name.
```
[the remark this answers is not in the document]
Comment (levkohimins) [reply cli/commands/terraform/creds/providers/provider.go:19]: Thanks.

File: cli/commands/terraform/creds/providers/provider.go [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+	Name() string
+	// GetCredentials returns a credentials.
```
[the remark this answers is not in the document]
Comment (levkohimins) [reply cli/commands/terraform/creds/providers/provider.go:21]: Thanks.

File: docs/_docs/04_reference/cli-options.md [hunk 24]
```diff
@@ -1306,3 +1306,23 @@ Specify the plan output directory for the `*-all` commands. Useful to save plans
...
+  "awsCredentials": {
+    "ACCESS_KEY_ID": "",
+    "SECRET_ACCESS_KEY": "",
+    "SESSION_TOKEN": ""
+  }
```
[the remark this answers is not in the document]
Comment (levkohimins) [reply docs/_docs/04_reference/cli-options.md:1329]: Improved the example.

File: docs/_docs/04_reference/cli-options.md [hunk 25]
```diff
@@ -1306,3 +1306,23 @@ Specify the plan output directory for the `*-all` commands. Useful to save plans
...
+}
+```
```
[the remark this answers is not in the document]
Comment (levkohimins) [reply docs/_docs/04_reference/cli-options.md:1331]: Excellent! I envy your ability to write good documentation. Thanks!

File: options/options.go [hunk 26]
```diff
@@ -297,6 +297,10 @@ type TerragruntOptions struct {
 
 	// Folder to store JSON representation of output files.
 	JsonOutputFolder string
+
+	// The path to the authentication command to obtain AWS credentials.
```
[the remark this answers is not in the document]
Comment (levkohimins) [reply options/options.go]: Fixed

File: cli/commands/flags.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (levkohimins) [reply cli/commands/flags.go]: Fixed

File: test/integration_test.go [hunk 27]
```diff
@@ -4123,6 +4124,28 @@ func TestReadTerragruntConfigIamRole(t *testing.T) {
...
 }
 
+func TestReadTerragruntAuthProviderCmd(t *testing.T) {
+	t.Parallel()
+
+	cleanupTerraformFolder(t, TEST_FIXTURE_AUTH_PROVIDER_CMD)
+	tmpEnvPath := copyEnvironment(t, TEST_FIXTURE_AUTH_PROVIDER_CMD)
+	rootPath := util.JoinPath(tmpEnvPath, TEST_FIXTURE_AUTH_PROVIDER_CMD)
+	appPath := util.JoinPath(rootPath, "app1")
+	mockPiplinesCmd := filepath.Join(rootPath, "mock-pipelines.sh")
```
[the remark this answers is not in the document]
Comment (levkohimins) [reply test/integration_test.go]: Renamed

File: test/integration_test.go (hunk 21 above)
Comment (levkohimins) [reply test/integration_test.go]: Renamed

File: docs/_docs/04_reference/cli-options.md [hunk 28]
```diff
@@ -1306,3 +1306,46 @@ Specify the plan output directory for the `*-all` commands. Useful to save plans
...
+
+```bash
+#!/usr/bin/env bash
+
+echo -n '{"envs": {"KEY": "a secret"}}'
+```
+
+You can use any technology you'd like, however, as long as Terragrunt can execute it. The expected pattern for using this flag is to populate the values dynamically using a secret store, etc.
+
+Note that more specific configurations (e.g. awsCredentials) take precedence over less specific configurations (e.g. envs).
```
Comment (yhakbar) [inline docs/_docs/04_reference/cli-options.md:1347]:
```suggestion
Note that more specific configurations (e.g. `awsCredentials`) take precedence over less specific configurations (e.g. `envs`).
```

NIT

File: docs/_docs/04_reference/cli-options.md [hunk 29]
```diff
@@ -1306,3 +1306,46 @@ Specify the plan output directory for the `*-all` commands. Useful to save plans
...
+
+If you would like to set credentials for AWS with this method, you are encouraged to use awsCredentials instead of envs, as these keys will be validated to conform to the officially supported environment variables expected by the AWS SDK.
+
+Other credential configurations will be supported in the future, but until then, if your provider authenticates via environment variables, you can use the envs field to fetch credentials dynamically from a secret store, etc before Terragrunt executes any IAC.
```
Comment (yhakbar) [inline docs/_docs/04_reference/cli-options.md:1351]:
```suggestion
Other credential configurations will be supported in the future, but until then, if your provider authenticates via environment variables, you can use the `envs` field to fetch credentials dynamically from a secret store, etc before Terragrunt executes any IAC.
```

File: docs/_docs/04_reference/cli-options.md [hunk 30]
```diff
@@ -1306,3 +1306,46 @@ Specify the plan output directory for the `*-all` commands. Useful to save plans
...
+#!/usr/bin/env bash
+
+echo -n '{"envs": {"KEY": "a secret"}}'
+```
+
+You can use any technology you'd like, however, as long as Terragrunt can execute it. The expected pattern for using this flag is to populate the values dynamically using a secret store, etc.
+
+Note that more specific configurations (e.g. awsCredentials) take precedence over less specific configurations (e.g. envs).
+
+If you would like to set credentials for AWS with this method, you are encouraged to use awsCredentials instead of envs, as these keys will be validated to conform to the officially supported environment variables expected by the AWS SDK.
```
Comment (yhakbar) [inline docs/_docs/04_reference/cli-options.md:1349]:
```suggestion
If you would like to set credentials for AWS with this method, you are encouraged to use `awsCredentials` instead of `envs`, as these keys will be validated to conform to the officially supported environment variables expected by the AWS SDK.
```

NIT

**Review by yhakbar — APPROVED** [review]
Looks great!

File: docs/_docs/04_reference/cli-options.md (hunk 28 above)
Comment (levkohimins) [reply docs/_docs/04_reference/cli-options.md]: Ah, Sorry for my inattention. Fixed everywhere.

**Review by yhakbar — APPROVED** [review]

**Review by denis256 — APPROVED** [review]

**Comment by levkohimins** [comment]
I am going to delete this branch and create a new one because _GitGuardian Security Checks_ found sensitive data in git history even though it was fake credentials.
